### DISCUSSION TITLE
"Müller Organ 1741" from hauptwerk.nl does not fully convert

### DISCUSSION THREAD

**Discussion by kwhite-uottawa** [GrandOrgue/OdfEdit#21] [General]
The free Müller Organ sample set from [https://www.hauptwerk.nl/bestellenen.php](https://www.hauptwerk.nl/bestellenen.php#) does not generate any connected ranks when converted to ODF with ODFEdit 2.3

Applying the attached patch to ODFEdit to supply default values instead of None for Action{Effect,Type}Code, and to ignore comment XML nodes during input parsing, gives a just about playable ODF (after some minor hand editing).

There are other minor graphical problems like text alignment on stop labels and missing elements on the Pictures and Wind panels.

[defnul.txt](https://github.com/GrandOrgue/ODFEdit/files/11480498/defnul.txt)

**Comment by eturpault** [comment]
It is interresting to analyze sample sets from new editors like this one, each editor having some ODF definition particularities, here the fact to not declare ActionTypeCode and ActionEffectCode in the StopRank objects, relying on their default values that I cannot know without having a sample set to deduce it, or the fact to have comments in the xml file or to set to 0 undefined labels clickable area dimensions, or the fact to have the panel Pictures without defined background image.
Thanks to this sample set I found a bug in OdfEdit about the label word wrap management, which generated wrong label TextRectWidth and TextRectHeight values.
Thank you for this interresting sample set and for the good proposed changes in the patch, I have included them in my next release.
Here is the ODF that I obtain with the various fixes done in OdfEdit (which will be present in the next release).
[Oosterwijtwerd11.zip](https://github.com/GrandOrgue/ODFEdit/files/11482428/Oosterwijtwerd11.zip)

**Reply by kwhite-uottawa** [reply]
Thanks!  The labels do indeed look much better positioned, and the speaking stops work!
However, the noise stops (KLOK, UURWERK, MOTOR) have gone.  (The reverse of with the 2.4 release: Where the noise stops were working and the speaking stops absent.)
As you mention, the Pictures panel and its organ and church switchable images is also missing.  Adding this isn't too hard to do manually anyway.

**Reply by eturpault** [reply]
The noise stops were missing because I generated yesterday the ODF with a development version of OdfEdit in which some features were disabled. Here is a new generated ODF. The Tremulant is not working because not yet managed by OdfEdit. The image panel is working, I made an implementation in OdfEdit to manage such a case.
[Oosterwijtwerd11.zip](https://github.com/GrandOrgue/ODFEdit/files/11492380/Oosterwijtwerd11.zip)

**Reply by kwhite-uottawa** [reply]
Everything now works as I would expect. Thanks!

**Reply by eturpault** [reply]
Here is a new ODF generated with OdfEdit v2.4 beta which now supports the tremulant convertion from HW to GO.
[Oosterwijtwerd11.zip](https://github.com/GrandOrgue/ODFEdit/files/11558582/Oosterwijtwerd11.zip)

**Reply by kwhite-uottawa** [reply]
Excellent.  Thanks again!
For those playing along, there is still manual editing required of the resulting GO file.  This is due to bugs in the HW file itself and *not* problems with the OdfEdit conversion!  (Staccato and portamento releases for "Octaaf 2 Bas" f#-b, and "Octaaf 2" f#-d' use the sample of the semitone below.  This makes for an interesting musical effect in fast scale passages...)
I'll try to provide a modified ODF file shortly.

**Reply by kwhite-uottawa** [reply]
Here's the modified file.  Release samples have been fixed.
[Oosterwijtwerd11R.zip](https://github.com/GrandOrgue/ODFEdit/files/11569295/Oosterwijtwerd11R.zip)

**Comment by kwhite-uottawa** [comment]
The Ott-Orgel [https://orgel.datzko.ch/](url) has a couple more edge cases for you.
The objects in the XML file are in lowercase: i.e. "_general" rather than "_General".  After converting them to mixed case, the file does convert except for the pull-down pedal.  It's defined as a "KeyAction" named "Pedal to Great" in the XML but it doesn't appear to be added to the ODF (perhaps it could be added as an always active invisible coupler?).
There are label errors too, but those are probably already fixed in your next release.

**Reply by eturpault** [reply]
This is an interresting use case, especially with all the objects types which are in lowercase, it is the first time I see this. I have added some lines of code in OdfEdit v2.4 beta to manage this case.

There is an error to fix manually in the XML (I downloaded the version for Hauptwerk 2), the last line repeats the closing tag </Hauptwerk> with a special character before, the XML parser that I use in OdfEdit doesn't like it, this last line has to be removed manually.

About the KeyAction "Pedal to Great" the problem is that in the HW ODF its state is not controlled by a switch object, so it is not possible to enable it (even by default), so OdfEdit doesn't convert it into a GO Coupler. The KeyAction object has no attribute to define in which state it must be the set by default.
All the other HW sample sets that I know have a switch to control each KeyAction object (for keyboards coupling control), even if the switch is engaged by default and not visible.

Several bitmaps of this sample set are taken from the installation package 000001 named HauptwerkStandardImages, especially to display the stops and the keys of the keyboard. This package is not present in the sample set (it is probably delivered by default in the HW application), so the stops and keys are rendered with default GO stops and keys bitmaps.

The labels are not rendered correctly because they are hidden by the drawstops buttons, since they are displayed at the same positions.
@oleg68 : should it be feasible in GrandOrgue that labels are displayed over and not under the drawstop buttons ?

**Reply by eturpault** [reply]
I made evolutions in OdfEdit v2.4 beta to manage the particularities of this sample set (I hope other sample sets have the same particularities as well, so these changes are not made for a single use case) : 
- if a KeyAction has no controlling switch and is linking different divisions or is making a keys shift, I build a GO Coupler which is engaged by default, which is not displayed and which links the divisions indicated by the KeyAction
- if a switch has no bitmap defined, I set the attribute DispLabelText of this switch with the name of the stop controlled by this switch.
It generates the following ODF which is not bad I think, what do you think ?

[Ott-Orgel.zip](https://github.com/GrandOrgue/ODFEdit/files/11593579/Ott-Orgel.zip)
It remains the problem of the labels which are hidden by the drawstop buttons, but it is no a problem since I set a label for the buttons.

**Comment by kwhite-uottawa** [comment]
The Ott-Orgel HW1 .rar and .exe files are corrupt, so getting the original HW1 file is a challenge.  Anyway,  for comparison, attached is the HW1 .organ file with a picture of the console after installing the ClerestorySSK font and the script I used to modify the Pipe999= locations.  No custom images were used, apparently.
[Ott-Orgel-HW1.zip](https://github.com/GrandOrgue/ODFEdit/files/11627874/Ott-Orgel-HW1.zip)

**Reply by eturpault** [reply]
Interresting. Personnally I prefer the console obtained by converting the HW2 ODF. With some manual changes done in the GO ODF generated by OdfEdit one can add this nice ClerestorySSK font for the stops and let GrandOrgue draw the pedal board with the default keys spacing.
